velox/docs/develop/view-and-writer-types.rst

```diff
@@ -0,0 +1,324 @@
+
+The object supports the following methods:
+
+- arg_type<E> value()      : unchecked access to the underlying value.
+
+- arg_type<E> operator \*() : unchecked access to the underlying value.
+
+- bool has_value()         : return true if the value is not null.
+
+- bool operator()          : return true if the value is not null.
+
+The nullity and the value accesses are decoupled, and hence if someone knows inputs are null-free,
+accessing the value does not have the overhead of checking the nullity. So is checking the nullity.
+Note that, unlike std::container, function calls to value() and operator* are r-values (temporaries) and not l-values,
+they can bind to const references and l-values but not references.
+
+OptionalAccessor<E> is assignable to and comparable with std::optional<arg_type<E>> for primitive types.
+The following expressions are valid, where array[0] is an optional accessor.
+
+.. code-block:: c++
+
+    std::optional<int> = array[0];
+    if(array[0] == std::nullopt) ...
+    if(std::nullopt == array[0]) ...
+    if(array[0]== std::optional<int>{1}) ...
+
+**NullableArrayView<T> and NullFreeArrayView<T>**
+
+NullableArrayView and NullFreeArrayView have interfaces similar to that of *std::vector<std::optional<V>>* and *std::vector<V>*,
+the code below shows the function arraySum, a range loop is used to iterate over the values.
+
+.. code-block:: c++
+
+  template <typename T>
+  struct ArraySum {
+    VELOX_DEFINE_FUNCTION_TYPES(T);
+
+    bool call(const int64_t& output, const arg_type<Array<int64_t>>& array) {
+      output = 0;
+      for(const auto& element : array) {
+        if (element.has_value()) {
+          output += element.value();
+        }
+      }
+      return true;
+    }
+  };
+
+
+ArrayView supports the following:
+
+- size_t size() : return the number of elements in the array.
+
+- operator[](size_t index) : access element at index. It returns either null_free_arg_type<T> or OptionalAccessor<T>.
+
+- ArrayView<T>::Iterator begin() : iterator to the first element.
+
+- ArrayView<T>::Iterator end() : iterator indicating end of iteration.
+
+- bool mayHaveNulls() : constant time check on the underlying vector nullity. When it returns false, there are definitely no nulls, a true does not guarantee null existence.
+
+- ArrayView<T>::SkipNullsContainer SkipNulls() : return an iterable container that provides direct access to non-null values in the underlying array. For example, the function above can be written as:
+
+.. code-block:: c++
+
+  template <typename T>
+  struct ArraySum {
+    VELOX_DEFINE_FUNCTION_TYPES(T);
+
+    bool call(const int64_t& output, const arg_type<Array<int64_t>>& array) {
+      output = 0;
+      for (const auto& value : array.skipNulls()) {
+        output += value;
+      }
+      return true;
+    }
+  };
+
+The skipNulls iterator will check the nullity at each index and skip nulls, a more performant implementation
```

Comment:
FYI, This is very neat, but in most cases, with default nullity if a null is present in the array or any complex input, dont presto functions return null ?

Reply:
No, I don't think so. Default null behavior makes the function automatically return null only if the top-level of the input is null. But when you iterate through elements inside an array or map, this skipNulls() allow not iterating over null elements.